src/istore_agg.c

```diff
@@ -59,11 +62,261 @@ state_extend(ISAggState *state, int n)
     if (state->size < state->used + n)
     {
         state->size = state->size * 2 > state->used + n ? state->size * 2 : state->used + n;
-        state       = repalloc(state, sizeof(ISAggState) + state->size * sizeof(BigIStorePair));
+        state       = repalloc(state,
+            offsetof(ISAggState, pairs) + state->size * sizeof(BigIStorePair));
     }
     return state;
 }
 
+/* for debug */
+#ifdef __GNUC__
+__attribute__((unused))
+#endif
+static void
+print_aggstate(ISAggState *state)
+{
+    for (int i=0; i < state->used; i++)
+        fprintf(stderr, "%d-%lu, ", state->pairs[i].key, state->pairs[i].val);
+    fprintf(stderr, "\n");
+}
+
+/* Internal SUM_FLOOR aggregation function for istore */
+static ISAggState *
+istore_agg_sum_floor_internal(ISAggState *state, IStore *istore)
+{
```

Comment:
That should share the bulk of the code with istore_agg_sum_internal, is it possible to move the common fragments into the helper function?

The bigistore/istore variants of that function are also almost identical. Perhaps, as a different matter, there is a way of introducing a helper function that receives the type of the pairs and the end istore (big or regular) and uses some macros to define variables of the right type (the only place where it matters seems to be the definitions and FIRST_PAIR, which already accepts the type as an argument).

Reply:
Yeah, macros should solve the issue, but it's harder to maintain them.

Reply:
I think I will add macro for addition with overflow check, it's quite simple and used in three places.